### PR TITLE
Add new struct in AcoountTransactionContext for V3.

### PR DESCRIPTION
**Pull request by NirLevi-starkware**
A few things:
1) A suggestion for a better name than CurrentAccountTransactionContext.
2) I saw one (not in test file) of too_many_arguments, is it a good solution here? or is there a better solution for this case?
3) When implementing the function get_account_tx_context is it better to implement match over version? asking in this case because in the V3 part I need to convert to V3 type to access the fields. 
4) Or maybe we want to implement (as in invoke in starknet_api) with default values for the version for which the fields not exist?
5) max_fee V3 - currently with value as in the rest of our code and with TODO. in the following PR will be fixed, needs access to block_context.gas_price, so we need to choose a solution (e.g. max_fee will receive as input the gas_price).

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)

### CODE REVIEW

**Review by elintul — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @dorimedini-starkware, @giladchase, and @NirLevi-starkware)

___
*[`crates/blockifier/src/transaction/objects.rs` line 188 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NfrAXMpA82r_uM_IpQT:-NfrAXMqBW9PLSQDt2TR:b-9wjl98) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L188)):*
> ```XML
> 
>     pub fn resource_bounds(&self) -> ResourceBoundsMapping {
>         self.resource_bounds.clone()
> ```

Perhaps we can have a cloning version of the macro?

**Review by giladchase — CHANGES_REQUESTED** [review]
> -   A suggestion for a better name than CurrentAccountTransactionContext.

It's fine IMO

> -   I saw one (not in test file) of too\_many\_arguments, is it a good solution here? or is there a better solution for this case?

In tests it's OK, in production only as a last resort IMO, it's a warning against a possible "god object".
Here, since `Current*` is a superset of `Deprecated*`, how about making `Current` hold an instance of `Deprecated` as a field? 
(If you do so, document why deprecated instance is there, it is somewhat unordinary)

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965#-:-NfqNY-d2HhFkS7Nii19:buf0i28)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @dorimedini-starkware and @NirLevi-starkware)

___
*[`crates/blockifier/src/transaction/objects.rs` line 128 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NfqOdsZDELLp_PBg28A:-NfqOds_EMSv8FUinbpV:b-9mwyxa) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L128)):*
> ```XML
>         })*
>     };
> }
> ```

1.  This is the same macro as the deprecated one, just rename the first one with something genric (maybe `add_getters`?) and use it for both.
2. Why do we need these getters? what is the invariant on the fields which forces them to be private? is it just for consistency with the cloned fields?
3. 



_Code quote:_
```XML
macro_rules! implement_inner_current_account_tx_context_getter_calls {
    ($(($field:ident, $field_type:ty)),*) => {
        $(pub fn $field(&self) -> $field_type {
            self.$field
        })*
    };
}
```
___
*[`crates/blockifier/src/transaction/objects.rs` line 130 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NfqOt4n-AhWKDnEUr9C:-NfqOt4n-AhWKDnEUr9D:bg2xf4g) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L130)):*
> ```XML
> }
> #[derive(Clone, Debug, Eq, PartialEq)]
> pub struct CurrentAccountTransactionContext {
> ```

We generally order things by "importance" inside a module.
Therefore the struct (and its methods) should be moved above the deprecated ones.

**Review by giladchase — COMMENTED** [review]
A better solution would be to break up `Current*` into smaller more targeted context objects (surely not all users of the context object need _all_ of its fields...), but that'll be kind of complicated to do.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965#-:-Nft7VJEAG4W5kSKw_s7:bup3lan)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @dorimedini-starkware and @NirLevi-starkware)

**Review by giladchase — COMMENTED** [review]
> When implementing the function get_account_tx_context is it better to implement match over version? asking in this case because in the V3 part I need to convert to V3 type to access the fields.

Most of time match is preferred, but not here IMO.
Note the duplication in the instantiation of both types: if we compose an instance of Deprecated inside current it'll allow you to remove the first 'if'. 

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965#-:-Nft9stBEMkhR20jf6ED:bh528nb)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @dorimedini-starkware and @NirLevi-starkware)

___
*[`crates/blockifier/src/transaction/objects.rs` line 129 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-Nft9TbJ9TYQtp_1aeW9:-Nft9TbJ9TYQtp_1aeWA:b-2u4rj4) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L129)):*
> ```XML
>     };
> }
> #[derive(Clone, Debug, Eq, PartialEq)]
> ```

Why no Default?

**Comment by NirLevi-starkware** [comment]
*[`crates/blockifier/src/transaction/objects.rs` line 129 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-Nft9TbJ9TYQtp_1aeW9:-NftT_oe4f8dD7RGbRA3:b-zff4cj) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L129)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Why no Default?
</blockquote></details>

`DataAvailabilityMode` does not implement `Default`.

**Review by giladchase — COMMENTED** [review]
> -   Or maybe we want to implement (as in invoke in starknet\_api) with default values for the version for which the fields not exist?

I don't think it's relevant here.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965#-:-NftWOzxC4HaaU8N4FkC:b-gkhkxy)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @dorimedini-starkware and @NirLevi-starkware)

**Comment by NirLevi-starkware** [comment]
> max\_fee V3 - currently with value as in the rest of our code and with TODO. in the following PR will be fixed, needs access to block\_context.gas\_price, so we need to choose a solution (e.g. max\_fee will receive as input the gas\_price).
Blocked by [This PR](https://reviewable.io/reviews/starkware-libs/blockifier/944#-).

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)* status: all files reviewed, 9 unresolved discussions (waiting on @NirLevi-starkware)

___
*[`crates/blockifier/src/transaction/objects.rs` line 143 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftWipo4_ZbWe8nZPn-:-NftWipo4_ZbWe8nZPn0:bo60uog) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L143)):*
> ```XML
>     paymaster_data: PaymasterData,
>     account_deployment_data: AccountDeploymentData,
> }
> ```

general question...
why is this context object needed?
it looks just like an invoke v3 struct with some extra data, why another similar object?
___
*[`crates/blockifier/src/transaction/objects.rs` line 145 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftWx9f-z5U3XUk5RTF:-NftWx9f-z5U3XUk5RTG:b-dtei1m) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L145)):*
> ```XML
> }
> 
> #[allow(clippy::too_many_arguments)]
> ```

prime candidate for using an external library to get a `derive` that auto-generates this `new` function, WDYT?
___
*[`crates/blockifier/src/transaction/transactions.rs` line 178 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftX8oX1MDHQpj4Qk0x:-NftX8oY6kBsGoVojdz4:bf8vztt) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/transactions.rs#L178)):*
> ```XML
>         if self.tx.version() == TransactionVersion::ZERO
>             || self.tx.version() == TransactionVersion::ONE
>             || self.tx.version() == TransactionVersion::TWO
> ```

if you're going with `if` and not `match`, use `<=`



_Suggestion:_
```XML
        if self.tx.version() <= TransactionVersion::TWO
```
___
*[`crates/blockifier/src/transaction/transactions.rs` line 188 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftXOme8hoTz7hClSTa:-NftXOme8hoTz7hClSTb:b-1im6xt) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/transactions.rs#L188)):*
> ```XML
>                     self.tx.nonce(),
>                     self.tx.sender_address(),
>                 ),
> ```

I agree with the above comments - better to create an intermediate object with this shared data, create it at the begining of this function, and then plug it in one of two places, depending on tx version



_Code quote:_
```XML
                    self.tx_hash(),
                    self.max_fee(),
                    self.tx.version(),
                    self.tx.signature(),
                    self.tx.nonce(),
                    self.tx.sender_address(),
                ),
```
___
*[`crates/blockifier/src/transaction/transactions.rs` line 289 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftXp1ZD9QaBlXjHAH8:-NftXp1ZD9QaBlXjHAH9:b-d626qs) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/transactions.rs#L289)):*
> ```XML
>     }
> 
>     pub fn get_account_tx_context(&self) -> AccountTransactionContext {
> ```

I don't like how this function is duplicated 3 times... though they are not exactly the same, there is a lot of common boilerplate.
Let's start with creating the intermediate struct containing: tx_hash, max_fee, version, signature, nonce and contract_address; maybe then you can implement a `construct_common_fields_struct(&self)` function on `AccountTransaction`, *once*, and then use it in the three variants of `get_account_tx_context`.
WDYT?



_Code quote:_
```XML
get_account_tx_context
```
___
*[`crates/blockifier/src/transaction/transactions.rs` line 380 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftYV6V5afkG44sewGX:-NftYV6V5afkG44sewGY:undefined) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/transactions.rs#L380)):*
> ```XML
>     pub fn get_account_tx_context(&self) -> AccountTransactionContext {
>         if self.tx.version() == TransactionVersion::ZERO
>             || self.tx.version() == TransactionVersion::ONE
> ```





_Suggestion:_
```XML
        if self.tx.version() <= TransactionVersion::ONE
```

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)* status: all files reviewed, 9 unresolved discussions (waiting on @elintul and @NirLevi-starkware)

___
*[`crates/blockifier/src/transaction/objects.rs` line 188 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NfrAXMpA82r_uM_IpQT:-NftYGyiDxPRRoolqJZr:bxplm0z) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L188)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Perhaps we can have a cloning version of the macro?
</blockquote></details>

I'm not sure i follow why we need getters here at all if there aren't any invariants on the fields we want to protect :thinking: 
Also it's a context inst so it should probably be as easy to use as possible

**Comment by NirLevi-starkware** [comment]
> max\_fee V3 - currently with value as in the rest of our code and with TODO. in the following PR will be fixed, needs access to block\_context.gas\_price, so we need to choose a solution (e.g. max\_fee will receive as input the gas\_price).

Blocked by [This PR](https://reviewable.io/reviews/starkware-libs/blockifier/944#-)

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)* status: all files reviewed, 9 unresolved discussions (waiting on @dorimedini-starkware, @elintul, and @NirLevi-starkware)

___
*[`crates/blockifier/src/transaction/objects.rs` line 145 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftWx9f-z5U3XUk5RTF:-NftZO6mDY20LQgxyIIc:b-ef9ky) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L145)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

prime candidate for using an external library to get a `derive` that auto-generates this `new` function, WDYT?
</blockquote></details>

IMO this indicates that the fields should be public; if there are no invariants on any of the fields, then why can't we use the default constructor?

**Review by dorimedini-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)* status: all files reviewed, 9 unresolved discussions (waiting on @elintul, @giladchase, and @NirLevi-starkware)

___
*[`crates/blockifier/src/transaction/objects.rs` line 145 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftWx9f-z5U3XUk5RTF:-NftZtlk9Md7Da0B2RwQ:b93wwia) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L145)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

IMO this indicates that the fields should be public; if there are no invariants on any of the fields, then why can't we use the default constructor?
</blockquote></details>

that is also fine by me

**Review by elintul — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)* status: all files reviewed, 10 unresolved discussions (waiting on @dorimedini-starkware, @giladchase, and @NirLevi-starkware)

*[a discussion](https://reviewable.io/reviews/starkware-libs/blockifier/965#-Nfu1L747EailfyEY0pz:-Nfu1L747EailfyEY0q-:b-flz951) (no related file):*
Please only include `CurrentAccountTransactionContext` in this PR.

___
___
*[`crates/blockifier/src/transaction/objects.rs` line 143 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftWipo4_ZbWe8nZPn-:-Nfu0qJu1kqpeOExt7zZ:b47yck) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L143)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

general question...
why is this context object needed?
it looks just like an invoke v3 struct with some extra data, why another similar object?
</blockquote></details>

To correctly model data: there's no account data that includes `max_fee` and `paymaster_data`, for example.
___
*[`crates/blockifier/src/transaction/objects.rs` line 188 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NfrAXMpA82r_uM_IpQT:-Nfu19Ja9LHXwTKlg1_l:bhrj22e) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L188)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

I'm not sure i follow why we need getters here at all if there aren't any invariants on the fields we want to protect :thinking: 
Also it's a context inst so it should probably be as easy to use as possible
</blockquote></details>

Yeah, they can be `pub`lic.

**Review by NirLevi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)* status: 0 of 2 files reviewed, 10 unresolved discussions (waiting on @dorimedini-starkware, @elintul, and @giladchase)

___
*[`crates/blockifier/src/transaction/objects.rs` line 128 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NfqOdsZDELLp_PBg28A:-NftV6DcAGafaf-OKwNi:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L128)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

1.  This is the same macro as the deprecated one, just rename the first one with something genric (maybe `add_getters`?) and use it for both.
2. Why do we need these getters? what is the invariant on the fields which forces them to be private? is it just for consistency with the cloned fields?
3.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/objects.rs` line 130 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NfqOt4n-AhWKDnEUr9C:-NftV1hkDaMFLP0xvYcz:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L130)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

We generally order things by "importance" inside a module.
Therefore the struct (and its methods) should be moved above the deprecated ones.
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/objects.rs` line 145 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftWx9f-z5U3XUk5RTF:-NfuE3ryC6GQeuF0c4rY:bx93ulr) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L145)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

that is also fine by me
</blockquote></details>

As you discussed before, there is dependency between `max_fee` and `resource_bounds`, so for now i will remain private. (maybe for the next PR?)
___
*[`crates/blockifier/src/transaction/objects.rs` line 188 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NfrAXMpA82r_uM_IpQT:-NfuEbmR5e2Dgw9dBMNm:bq5uo8d) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/objects.rs#L188)):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Yeah, they can be `pub`lic.
</blockquote></details>

See the conversation above.
___
*[`crates/blockifier/src/transaction/transactions.rs` line 178 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftX8oX1MDHQpj4Qk0x:-NfuAeOUBc9sMzsi-J38:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/transactions.rs#L178)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

if you're going with `if` and not `match`, use `<=`
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/transactions.rs` line 188 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftXOme8hoTz7hClSTa:-NfuEmH1Bjn5GmXDH5Ta:b-l04pwl) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/transactions.rs#L188)):*
<details><summary><i>Previously, dorimedini-starkware wrote…</i></summary><blockquote>

I agree with the above comments - better to create an intermediate object with this shared data, create it at the begining of this function, and then plug it in one of two places, depending on tx version
</blockquote></details>

This intermediate object leads to changes in access to the fields in many places in the code. if we think that we should use an intermediate object, i will do it in different PR.
___
*[`crates/blockifier/src/transaction/transactions.rs` line 380 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftYV6V5afkG44sewGX:-NfuAqDX2luVHlOhEviU:b-896fix) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/transactions.rs#L380)):*
> ```XML
>     pub fn get_account_tx_context(&self) -> AccountTransactionContext {
>         if self.tx.version() == TransactionVersion::ZERO
>             || self.tx.version() == TransactionVersion::ONE
> ```

Done.

**Review by NirLevi-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)* status: 0 of 2 files reviewed, 10 unresolved discussions (waiting on @dorimedini-starkware, @elintul, and @giladchase)

*[a discussion](https://reviewable.io/reviews/starkware-libs/blockifier/965#-Nfu1L747EailfyEY0pz:-NfuG9OP5PlkrCsdzBGR:b-896fix) (no related file):*
<details><summary><i>Previously, elintul (Elin) wrote…</i></summary><blockquote>

Please only include `CurrentAccountTransactionContext` in this PR.
</blockquote></details>

Done.

___

**Review by dorimedini-starkware — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/blockifier/965)* status: all files reviewed, 6 unresolved discussions (waiting on @elintul, @giladchase, and @NirLevi-starkware)

___
*[`crates/blockifier/src/transaction/transactions.rs` line 188 at r1](https://reviewable.io/reviews/starkware-libs/blockifier/965#-NftXOme8hoTz7hClSTa:-NfuSUbf1kbOaSBgCA9O:bp0myw5) ([raw file](https://github.com/starkware-libs/blockifier/blob/8d8674840923e2d17b6442a329d36a7112ba9b2a/crates/blockifier/src/transaction/transactions.rs#L188)):*
<details><summary><i>Previously, NirLevi-starkware (Nir Levi) wrote…</i></summary><blockquote>

This intermediate object leads to changes in access to the fields in many places in the code. if we think that we should use an intermediate object, i will do it in different PR.
</blockquote></details>

set a TODO about it and I'll unblock

**Comment by NirLevi-starkware** [comment]
Replaced by #967 , #968 , #970 .

**Pull request closed without merging**
